### PR TITLE
Back to using copy.canvasItem. Pating now occurs at position x y on the canvas.

### PR DESCRIPTION
**Pull request by wdednam**
This change is [Reviewable](https://reviewable.io/reviews/highperformancecoder/minsky/117)

### CODE REVIEW

File: gui-tk/minsky.tcl [hunk 1]
```diff
@@ -703,6 +703,12 @@ proc cut {} {
     minsky.cut
 }
 
+# Paste item(s) at position x and y on the canvas upon click down. For ticket 1098
+proc pasteAt {} {
+	canvas.mouseDown [get_pointer_x .wiring.canvas] [get_pointer_y .wiring.canvas]
```
Comment (highperformancecoder) [inline gui-tk/minsky.tcl]: The problem with this approach is that the mouse is currently over the "Paste" menu item, particularly with the defintion on line 690

File: gui-tk/minsky.tcl [hunk 2]
```diff
@@ -786,8 +792,8 @@ bind . <$meta-x> {minsky.cut}
 bind . <$meta-X> {minsky.cut}
 bind . <$meta-c> {minsky.copy}
 bind . <$meta-C> {minsky.copy}
-bind . <$meta-v> {minsky.paste}
-bind . <$meta-V> {minsky.paste}
+bind . <$meta-v> {pasteAt}
+bind . <$meta-V> {pasteAt}
```
Comment (highperformancecoder) [inline gui-tk/minsky.tcl]: This would work...

File: model/minsky.cc [hunk 3]
```diff
@@ -265,6 +265,7 @@ namespace minsky
     canvas.setItemFocus(model->addGroup(g));
```
Comment (highperformancecoder) [inline model/minsky.cc]: OK for the moveTo to work on line 268, yes g needs to be added to model. But you can ditch the call to setItemFocus.

Comment (highperformancecoder) [reply model/minsky.cc]: Also solves the segfault problem in adjustWires when the two groups don't share an ancester (that should probably be fixed too, but not so urgent).

**Comment by highperformancecoder** [comment]
The problem I see with this approach is that it pastes the clipboard where the mouse pointer currently is - which will be wherever the menu cascade leaves it. My approach has been to "attach" the pasted items to the mouse, requiring a further click to indicate where the items should be positioned. This is a consistent protocol with the copy item operation and the insert item operation.

The alternative is to introduce a concept of "the cursor", which you'd have to move around with the mouse, and where copyitem, paste and insert (and incidentally text items) would place their payloads. This is the wordprocessor model.

I just checked what a drawing program might do, eg LibreOffice. It appears to paste over the top of the previously selected item. Ouch! I like my way better than that.

**Comment by wdednam** [comment]
Please don't merge yet. On copy of an exiting item or selection of items the new copy still ends up on top of the original selection after pasting, and then both remain in selection, not just the new copy, which one wants to move off the original. This is obviously not a problem when cutting and pasting...

File: model/minsky.cc [hunk 4]
```diff
@@ -255,27 +255,47 @@ namespace minsky
...
+    if (copyOfItems.size()%2==0) numPaste = copyOfItems.size()/2-g->numWires();
+    else numPaste = copyOfItems.size()/2-g->numWires()-1;    
     for (auto& i: copyOfItems)
-      {
-        model->addItem(i);
-        canvas.selection.toggleItemMembership(i);
-        assert(!i->ioVar());
+      {		
+		 // Ensure only a single copy of a selection of items exists at time 
+		 if (count<numPaste && model->uniqueItems())
```
Comment (highperformancecoder) [inline model/minsky.cc]: I don't follow at all what you're doing with numPaste, an uniqueItems is a quite expensive operation, so shouldn't be done in a loop like this.

File: model/minsky.cc [hunk 5]
```diff
@@ -255,27 +255,47 @@ namespace minsky
...
-        model->addItem(i);
-        canvas.selection.toggleItemMembership(i);
-        assert(!i->ioVar());
+      {		
+		 // Ensure only a single copy of a selection of items exists at time 
+		 if (count<numPaste && model->uniqueItems()) 
+		 {	 
+           canvas.model->addItem(i);			  
+           canvas.selection.ensureItemInserted(i);			 
+           canvas.setItemFocus(i);
```
Comment (highperformancecoder) [inline model/minsky.cc]: Why are we setting itemFocus here? The next iteration undoes it.

File: model/minsky.cc [hunk 6]
```diff
@@ -255,27 +255,47 @@ namespace minsky
 //      throw runtime_error("cannot save to "+fileName);
   }
 
-  void Minsky::paste(float x, float y)
```
Comment (highperformancecoder) [inline model/minsky.cc]: Aww - I quite liked the idea of passing in coordinates for where to paste the contents...

File: model/minsky.cc (hunk 4 above)
Comment (wdednam) [reply model/minsky.cc]: Thanks, your suggestion of clearing the selection first did the trick.

File: model/minsky.cc (hunk 5 above)
Comment (wdednam) [reply model/minsky.cc]: I tried to select just one of the items after the loop, via findAny:

    auto oneItem=g->findAny(&Group::items,[&](const ItemPtr& i)
                         {return false;});          
    if (oneItem) canvas.setItemFocus(oneItem);

This just pastes the copy of the selection on top of the original one and then I can't separate them.

I don't know if there is a better method?

File: model/minsky.cc (hunk 6 above)
Comment (wdednam) [reply model/minsky.cc]: I also thought it was a good idea, and think I came close to getting it right but feel my understanding of how the coordinates are handled between TCL and Minsky is still on the low side.

File: gui-tk/wiring.tcl [hunk 7]
```diff
@@ -629,7 +629,7 @@ proc contextMenu {x y X Y} {
                 renameVariableInstances
             }
             .wiring.context add command -label "Edit" -command "editItem"
-            .wiring.context add command -label "Copy" -command "canvas.copyItem"
+            .wiring.context add command -label "Copy" -command "copy.canvasItem"
```
Comment (highperformancecoder) [inline gui-tk/wiring.tcl]: Where is copy.canvasItem defined?

File: gui-tk/wiring.tcl [hunk 8]
```diff
@@ -650,7 +650,7 @@ proc contextMenu {x y X Y} {
                .wiring.context add command -label "Initialise Random" \
                     -command "initRandom" 
             }
-            .wiring.context add command -label "Copy" -command "canvas.copyItem"
+            .wiring.context add command -label "Copy" -command "copy.canvasItem"
```
Comment (highperformancecoder) [inline gui-tk/wiring.tcl]: Where is copy.canvasItem defined?

File: gui-tk/minsky.tcl [hunk 9]
```diff
@@ -703,6 +703,12 @@ proc cut {} {
     minsky.cut
 }
 
+# Paste item(s) at position x and y on the canvas upon click down. For ticket 1098
+proc pasteAt {} {
+	canvas.mouseDown [get_pointer_x .wiring.canvas] [get_pointer_y .wiring.canvas]
+	minsky.paste
```
Comment (highperformancecoder) [inline gui-tk/minsky.tcl]: Missing the new coordinate arguments

File: model/minsky.cc [hunk 10]
```diff
@@ -255,7 +255,7 @@ namespace minsky
 //      throw runtime_error("cannot save to "+fileName);
   }
 
-  void Minsky::paste()
+  void Minsky::paste(float x, float y)
```
Comment (highperformancecoder) [inline model/minsky.cc]: Probably also need a boolean argument to indicate that paste occurred from the edit menu. If so, then the newly pasted items need to be left in the selection, and one item of the selection needs to be left in itemFocus (effectively attaching the selection to the mouse tip). If pasted from ^V or the context menu, the itemFocus should be cleared, and I'm not sure about the selection.

**Comment by highperformancecoder** [comment]
On Tue, Jan 07, 2020 at 04:27:38AM -0800, Wynand wrote:
> 
> I tried to select just one of the items after the loop, via findAny:
> 
> auto oneItem=g->findAny(&Group::items,[&](const ItemPtr& i)
>                      {return false;});
> if (oneItem) canvas.setItemFocus(oneItem);
> 
> This just pastes the copy of the selection on top of the original one and then
> I can't separate them.
> 
> I don't know if there is a better method?
> 

Instead of using findAny, which in this case will not return anything
since you are returning the constant false, just check if g->items is
empty, and if not, just grab the first element (g->items[0]).

Cheers

-- 

----------------------------------------------------------------------------
Dr Russell Standish                    Phone 0425 253119 (mobile)
Principal, High Performance Coders     hpcoder@hpcoders.com.au
                      http://www.hpcoders.com.au
----------------------------------------------------------------------------

File: gui-tk/wiring.tcl (hunk 7 above)
Comment (wdednam) [reply gui-tk/wiring.tcl]: Thanks for catching this. It must have occurred when I tested minsky.paste here instead of copy.canvasItem which was originally there.

File: gui-tk/minsky.tcl (hunk 9 above)
Comment (highperformancecoder) [reply gui-tk/minsky.tcl]: It seems to me you still want pasteAt for ^V handling.

Comment (wdednam) [reply gui-tk/minsky.tcl]: Why is that there? I don't understand. They don't show up in my local bugfix-1098 branch?? I have also added the bugfix-901 changes because you have just merged that PR into your master branch. maybe I should revert and git pull your master into my local bugfix-1098 again? It has become a bit messy...

File: test/00/t0036a.sh [hunk 11]
```diff
@@ -42,7 +42,8 @@ proc afterMinskyStarted {} {uplevel #0 {
   canvas.mouseUp [expr [\$item.x]+0.55*[\$item.width]] [expr [\$item.y]-0.55*[\$item.height]] 
   assert {[canvas.selection.groups.size]==1}
   newSystem
-  paste
+  # For ticket 1098. Paste items at mouse position
+  paste [expr [\$item.x]+0.55*[\$item.width]] [expr [\$item.y]-0.55*[\$item.height]]
```
Comment (highperformancecoder) [inline test/00/t0036a.sh]: Should revert this change too
